### ISSUE TITLE
Mapgen Format Question

### ISSUE THREAD

**Issue by vache** [CleverRaven/Cataclysm-DDA#13564]
How do you feel about mapgen JSONs that look like this: http://pastebin.com/eJtf68Mj

I'm working on a map editor and the JSON writer that it uses wants to output everything in alphabetical order and in fully indented style.  It is theoretically possible for me to split it up into lots of smaller pieces to get it the way we are used to, but I predict it will be a huge amount of work, and probably one of the last things I tackle.  I wanted to get some opinions here before I make my first real release of it, because I don't want people to make a bunch of buildings that have no hope of ever getting mainlined.  The main difficulty I foresee from it is that it doesn't really match our existing style, and the spread out format could make it hard to diagnose problems.

**Comment by Coolthulhu** [comment]
Does it have to reorder the fields?
You could have a predefined array defining the sorting order and use that instead of pure alphabetical order, only falling back to alphabetical for stuff that isn't in the array (putting it after the known stuff).

As for the full indentation: would it be possible to measure the length of resulting object? You could then have it keep short objects (x/y arrays, place_monsters etc.) on one line.

If it is to output huge unreadable jsons, it would need to be able to read them too and later automatically reformat them to something more readable.

**Comment by vache** [comment]
The alphabetical ordering is automatically enforced, and indented/compact is an all or nothing deal.  The solution for both is to split them up into smaller units and then stitch it all together at the end, but this isn't a problem I would prefer to tackle at this time, at least not until I can support all (or almost all) features that can be added in.

My intent is for the final version to be able to read and write JSONs, and it should be able to read both formats naturally (though there are some features in handmade JSON that may not be supported at first) and then it's just a matter of solving the first problem.

**Comment by kevingranade** [comment]
Yea that's not the nicest thing ever as far as readability, but it's mostly intended to be write-once I suspect, and if it makes mapgen significantly easier (i.e. more people contributing nice mapgen elements) it's worthwhile to ease up on the formatting a bit.  Maybe we can find or make a json formatter some time :P

**Comment by i2amroy** [comment]
Yeah, only big problem I see with it is the same things that have already been mentioned, notably that the ordering makes it very difficult to understand what it actually looks like by looking at the JSON.

I love the idea of this, but honestly I'd really like to see at least some readability improvements (it would be very nice to have "rows", "terrain", and "furniture" at the very top, at least) before I would wholeheartedly suggest it's widespread inclusion into the JSON files.

>Maybe we can find or make a json formatter some time :P

We went looking for one of these online a while back IIRC, and the basic answer came back as "make one yourself". :P

**Comment by kevingranade** [comment]
Just to be sure this is clear, this would be for new mapgen entries, not
reformatting the existing ones.  I'd be against the latter.
What it boils down to for me is that mapgen in particular is a difficult
enough format to work with that if the question is between 'hard to read
json that you can use a WYSIWYG editor on' and 'easy to read json (sort
of...) that you have to edit manually', I lean toward the former.

**Comment by mugling** [comment]
Closed as OP question was answered
